p2p/src/net/libp2p/mod.rs

```diff
@@ -179,21 +197,65 @@ where
         .collect::<Vec<net::AddrInfo<T>>>()
 }
 
+impl<T> TryInto<net::PeerInfo<T>> for IdentifyInfo
+where
+    T: NetworkService<PeerId = PeerId, ProtocolId = String>,
+{
+    type Error = P2pError;
+
+    fn try_into(self) -> Result<net::PeerInfo<T>, Self::Error> {
+        let proto = self.protocol_version.clone();
+        let (version, net) =
+            match sscanf::scanf!(proto, "/{}/{}.{}.{}-{}", String, u8, u8, u16, String) {
+                Err(e) => Err(P2pError::ProtocolError(ProtocolError::InvalidProtocol)),
+                Ok((proto, maj, min, pat, magic)) => {
+                    if proto != "mintlayer" {
+                        return Err(P2pError::ProtocolError(ProtocolError::InvalidProtocol));
+                    }
+
+                    // TODO: `impl TryInto<ChainType> for u32`?
+                    let net = match magic.as_str() {
+                        "1a64e5f1" => common::chain::config::ChainType::Mainnet,
+                        _ => return Err(P2pError::ProtocolError(ProtocolError::UnknownNetwork)),
+                    };
+
+                    Ok((common::primitives::version::SemVer::new(maj, min, pat), net))
+                }
+            }?;
+
+        Ok(net::PeerInfo {
+            peer_id: PeerId::from_public_key(&self.public_key),
+            net,
+            version,
+            agent: Some(self.agent_version),
+            protocols: self.protocols,
+        })
+    }
+}
+
 #[async_trait]
 impl NetworkService for Libp2pService {
     type Address = Multiaddr;
-    type Socket = Libp2pSocket;
     type Strategy = Libp2pStrategy;
     type PeerId = PeerId;
+    type ProtocolId = String;
+    type RequestId = RequestId;
+    type MessageId = MessageId;
     type ConnectivityHandle = Libp2pConnectivityHandle<Self>;
-    type FloodsubHandle = Libp2pFloodsubHandle<Self>;
+    type PubSubHandle = Libp2pPubSubHandle<Self>;
+    type SyncingHandle = Libp2pSyncHandle<Self>;
 
     async fn start(
         addr: Self::Address,
         strategies: &[Self::Strategy],
-        topics: &[FloodsubTopic],
+        topics: &[PubSubTopic],
+        config: Arc<common::chain::ChainConfig>,
         timeout: std::time::Duration,
-    ) -> error::Result<(Self::ConnectivityHandle, Self::FloodsubHandle)> {
+    ) -> error::Result<(
+        Self::ConnectivityHandle,
+        Self::PubSubHandle,
+        Self::SyncingHandle,
+    )> {
```

Comment:
P2P consists of three separate services (swarm maintenance, syncing, and block/tx publish/subscribe). Each of these services take one of these handles and start processing any networks events relating to that specific handle. The services are started right after the network service provider itself has been created so having them in a struct IMO doesn't add much value because the struct is discarded almost immediately after the function returns.

On a related note though, when you have the time, I'd like to have a chat about any ideas on how we could bootstrap P2P as it's not one service but three services (so three subsystems) and they depend on each other, e.g., `SwarmManager` informs `SyncManager` about incoming/closed connections and `SyncManager` notifies `SwarmManager` if a peer has published invalid data. I found it somewhat challenging to start all these services since they require handles to each other but those handles are acquired only after they have been started.